contracts/handlers/XC20Handler.sol

```diff
@@ -0,0 +1,106 @@
+        @param data Consists of {amount} padded to 32 bytes.
+        @notice Data passed into the function should be constructed as follows:
+        amount                                      uint256     bytes   0 - 32
+        destinationRecipientAddress     length      uint256     bytes  32 - 64
+        destinationRecipientAddress                 bytes       bytes  64 - END
+        @dev Depending if the corresponding {tokenAddress} for the parsed {resourceID} is
+        marked true in {_burnList}, deposited tokens will be burned, if not, they will be locked.
+        @return an empty data.
+     */
+    function deposit(
+        bytes32 resourceID,
+        address depositor,
+        bytes   calldata data
+    ) external override onlyBridge returns (bytes memory) {
+        uint256        amount;
+        (amount) = abi.decode(data, (uint));
+
+        address tokenAddress = _resourceIDToTokenContractAddress[resourceID];
+        require(_contractWhitelist[tokenAddress], "provided tokenAddress is not whitelisted");
+
+        if (_burnList[tokenAddress]) {
+            burnERC20(tokenAddress, depositor, amount);
+        } else {
+            lockERC20(tokenAddress, depositor, address(this), amount);
+        }
+    }
+
+    /**
+        @notice Proposal execution should be initiated when a proposal is finalized in the Bridge contract.
+        by a relayer on the deposit's destination chain.
+        @param resourceID ResourceID to be used when making deposits.
+        @param data Consists of {resourceID}, {amount}, {lenDestinationRecipientAddress},
+        and {destinationRecipientAddress} all padded to 32 bytes.
+        @notice Data passed into the function should be constructed as follows:
+        amount                                 uint256     bytes  0 - 32
+        destinationRecipientAddress length     uint256     bytes  32 - 64
+        destinationRecipientAddress            bytes       bytes  64 - END
+     */
+    function executeProposal(bytes32 resourceID, bytes calldata data) external override onlyBridge {
+        uint256       amount;
+        uint256       lenDestinationRecipientAddress;
+        bytes  memory destinationRecipientAddress;
+
+        (amount, lenDestinationRecipientAddress) = abi.decode(data, (uint, uint));
+        destinationRecipientAddress = bytes(data[64:64 + lenDestinationRecipientAddress]);
+
+        bytes20 recipientAddress;
+        address tokenAddress = _resourceIDToTokenContractAddress[resourceID];
+
+        assembly {
+            recipientAddress := mload(add(destinationRecipientAddress, 0x20))
+        }
+
+        require(_contractWhitelist[tokenAddress], "provided tokenAddress is not whitelisted");
+
+        if (_burnList[tokenAddress]) {
+            mintERC20(tokenAddress, address(recipientAddress), amount);
+        } else {
+            releaseERC20(tokenAddress, address(recipientAddress), amount);
```

Comment:
for lock and release, I can see you are trying to reuse them from ERC20Safe, however, from what I have tested, I doubt our `_safeTransfer` and `_safeTransferFrom` would work in this case, basically Astar xc20 precompile only works with whatever method is implemented in their rust codebase, you could start to test them out and see, maybe I tested it wrong, also [there](https://github.com/AstarNetwork/astar-frame/blob/polkadot-v0.9.28/precompiles/assets-erc20/src/lib.rs) you can fine their rust XC20 impl